src/Microsoft.Health.Fhir.Core/Features/Operations/DataConvert/DataConvertEngine.cs

```diff
@@ -0,0 +1,150 @@
+            catch (ImageFetchException fetchException)
+            {
+                _logger.LogError(fetchException, "Failed to fetch the templates from remote.");
+                throw new GetTemplateSetFailedException(string.Format(Resources.GetTemplateSetFailed, fetchException.Message), fetchException);
+            }
+            catch (ImageValidationException validationException)
+            {
+                _logger.LogError(validationException, "Failed to validate the downloaded image.");
+                throw new GetTemplateSetFailedException(string.Format(Resources.GetTemplateSetFailed, validationException.Message), validationException);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Unhandled exception: failed to get template set.");
+                throw new GetTemplateSetFailedException(string.Format(Resources.GetTemplateSetFailed, ex.Message), ex);
+            }
+
+            var dataConverter = _dataConverterMap.GetValueOrDefault(convertRequest.InputDataType);
+            if (dataConverter == null)
+            {
+                // This case should never happen.
+                _logger.LogError("Invalid input data type for conversion.");
+                throw new RequestNotValidException("Invalid input data type for conversion.");
+            }
+
+            try
+            {
+                string bundleResult = dataConverter.Convert(convertRequest.InputData, convertRequest.EntryPointTemplate, templateProvider);
+                return new DataConvertResponse(bundleResult);
+            }
+            catch (DataParseException dpe)
+            {
+                _logger.LogError(dpe, "Unable to parse the input data.");
+                throw new InputDataParseErrorException(string.Format(Resources.InputDataParseError, convertRequest.InputDataType.ToString()), dpe);
+            }
+            catch (InitializeException ie)
+            {
+                _logger.LogError(ie, "Fail to initialize the convert engine.");
+                throw new ConvertEngineInitializeException(Resources.ConvertEngineInitializeFailed, ie);
+            }
+            catch (FhirConverterException fce)
+            {
+                _logger.LogError(fce, "Data convert process failed.");
+                throw new DataConvertFailedException(string.Format(Resources.DataConvertFailed, fce.Message), fce);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Unhandled exception: data convert process failed.");
+                throw new DataConvertFailedException(string.Format(Resources.DataConvertFailed, ex.Message), ex);
+            }
+        }
+
+        private string ExtractRegistryServer(string templateSetReference)
+        {
+            var referenceComponents = templateSetReference.Split(ImageRegistryDelimiter);
+            if (referenceComponents.Length <= 1 || string.IsNullOrWhiteSpace(referenceComponents.First()))
```

Comment:
>referenceComponents.Length <= 1 [](start = 16, length = 31)

Is it ok if it's more than 2 components?
And why parsing happens here and not during `DataConvertRequest` creation?
Wouldn't it be better to do validation of request inside request itself, rather than later?

Reply:
Yes. The format is OK. 
I have moved the format validation to the controller. Thanks